File: internal/brand/brand_test.go
package brand

import (
	"strings"
	"testing"
)

func TestEmbeddedSchemaVersion(t *testing.T) {
	if got := Load().SchemaVersion; got != SupportedSchemaVersion {
		t.Fatalf("embedded schemaVersion = %d, want %d — regenerate with scripts/sync-branding.mjs", got, SupportedSchemaVersion)
	}
}

func TestNameAndShortName(t *testing.T) {
	cfg := Load()
	for _, tc := range []struct {
		locale, name, short string
	}{
		{"zh-CN", "布丁盒子", "布丁盒子"},
		{"en-US", "Pudding Box", "Pudding"},
	} {
		if got := cfg.Name(tc.locale); got != tc.name {
			t.Errorf("Name(%q) = %q, want %q", tc.locale, got, tc.name)
		}
		if got := cfg.ShortName(tc.locale); got != tc.short {
			t.Errorf("ShortName(%q) = %q, want %q", tc.locale, got, tc.short)
		}
	}
}

// The desktop shell resolves the UI language to a bare "zh"/"en" (see
// cmd/octo-desktop/lang.go resolveLang), so the bare subtag must find zh-CN.
func TestBareLanguageSubtagResolves(t *testing.T) {
	cfg := Load()
	if got := cfg.Name("zh"); got != "布丁盒子" {
		t.Errorf(`Name("zh") = %q, want 布丁盒子`, got)
	}
	if got := cfg.Name("en"); got != "Pudding Box" {
		t.Errorf(`Name("en") = %q, want "Pudding Box"`, got)
	}
	if got := cfg.ShortName("en"); got != "Pudding" {
		t.Errorf(`ShortName("en") = %q, want "Pudding"`, got)
	}
}

// Traditional Chinese falls back to the Simplified value through the shared
// language subtag. This phase ships no zh-TW entry on purpose (the product
// requirement routes Traditional systems to English at the UI layer), so this
// test pins the *mechanism*, not a promise of zh-TW support.
func TestUnlistedLocaleFallsBack(t *testing.T) {
	cfg := Load()
	if got := cfg.Name("zh-TW"); got != "布丁盒子" {
		t.Errorf(`Name("zh-TW") = %q, want the zh-CN value`, got)
	}
	if got := cfg.Name("de-DE"); got != "Pudding Box" {
		t.Errorf(`Name("de-DE") = %q, want the en-US default`, got)
	}
	if got := cfg.Name(""); got == "" {
		t.Error(`Name("") returned empty; an unknown locale must still yield a name`)
	}
}

func TestCopyrightForSubstitutesYear(t *testing.T) {
	cfg := Load()
	if raw := cfg.Copyright("zh-CN"); !strings.Contains(raw, "{year}") {
		t.Fatalf("Copyright(zh-CN) = %q, expected an unsubstituted {year} placeholder", raw)
	}
	for _, locale := range []string{"zh-CN", "en-US"} {
		got := cfg.CopyrightFor(locale, 2026)
		if strings.Contains(got, "{year}") {
			t.Errorf("CopyrightFor(%q) left the placeholder in: %q", locale, got)
		}
		if !strings.Contains(got, "2026") {
			t.Errorf("CopyrightFor(%q) = %q, want it to contain 2026", locale, got)
		}
	}
}

// Class B: one fixed value per field, deliberately not localized. The Chinese
// file description is a fixed display string, not an identifier, so it stays
// Chinese on an English Windows too.
func TestWindowsDisplayValuesAreFixed(t *testing.T) {
	cfg := Load()
	if got := cfg.Display(DisplayWindows, DisplayProductName); got != "Pudding Box" {
		t.Errorf("Display product name = %q, want %q", got, "Pudding Box")
	}
	if got := cfg.Display(DisplayWindows, DisplayWindowsDescription); got != "布丁盒子" {
		t.Errorf("Display file description = %q, want %q", got, "布丁盒子")
	}
}

// Every declared identifier constant must resolve, otherwise a call site gets
// an empty string where a path or a lock name belongs.
func TestDeclaredIdentifiersResolve(t *testing.T) {
	cfg := Load()
	for _, key := range []string{
		IdentifierExeName, IdentifierCLIExeName, IdentifierPortableDirName,
		IdentifierSingleInstanceID, IdentifierWindowsInternalName,
		IdentifierDataRoot, IdentifierWorkspaceDir, IdentifierPort,
		IdentifierCLICommand, IdentifierConfigDir, IdentifierEnvPrefix,
		IdentifierUIProtocol, IdentifierPairingProtocol, IdentifierGoModule,
		IdentifierMacBundleID, IdentifierInnoAppID, IdentifierMobileAppID,
		IdentifierInstallDir, IdentifierLinuxDesktopEntry,
	} {
		if cfg.Identifier(key) == "" {
			t.Errorf("Identifier(%q) is empty", key)
		}
	}
}

// Class C values are machine-read, so they must be printable ASCII without
// spaces on every platform. scripts/brand-schema.mjs enforces this on the
// source; this asserts the embedded copy the Go binary actually ships with.
func TestIdentifiersAreAsciiWithoutSpaces(t *testing.T) {
	cfg := Load()
	groups := map[string]map[string]string{
		"identifiers.current": cfg.Identifiers.Current,
		"identifiers.future":  cfg.Identifiers.Future,
		"visual.logo":         cfg.Visual.Logo,
		// visual.colors is absent until the brand colour is agreed; a nil map
		// here simply contributes nothing rather than failing.
		"visual.colors": cfg.Visual.Colors,
	}
	for group, links := range cfg.Links {
		groups["links."+group] = links
	}
	for group, values := range groups {
		for key, value := range values {
			for _, r := range value {
				if r < '!' || r > '~' {
					t.Errorf("%s.%s = %q contains %q; class C must be printable ASCII without spaces", group, key, value, r)
					break
				}
			}
		}
	}
}

// The single-instance lock must not collide with an Octo install on the same
// machine — that is the whole reason the identifier changes this phase.
func TestSingleInstanceIDIsRebranded(t *testing.T) {
	got := Load().Identifier(IdentifierSingleInstanceID)
	if got != "app.puddingbox.desktop" {
		t.Errorf("single instance id = %q, want app.puddingbox.desktop", got)
	}
	if strings.Contains(got, "octo") {
		t.Errorf("single instance id %q still references the upstream app", got)
	}
}

// Compatibility identifiers stay on the upstream values this phase; changing
// them breaks in-place upgrades for already-installed copies.
func TestCompatibilityIdentifiersUnchanged(t *testing.T) {
	cfg := Load()
	for key, want := range map[string]string{
		IdentifierCLICommand:  "octo",
		IdentifierConfigDir:   "~/.octo",
		IdentifierEnvPrefix:   "OCTO_",
		IdentifierGoModule:    "github.com/open-octo/octo-agent",
		IdentifierMacBundleID: "dev.octo-agent.desktop",
		IdentifierMobileAppID: "dev.octo.mobile",
	} {
		if got := cfg.Identifier(key); got != want {
			t.Errorf("Identifier(%q) = %q, want %q (must stay on the upstream value this phase)", key, got, want)
		}
	}
}

func TestLinksAndAssets(t *testing.T) {
	cfg := Load()
	if got := cfg.Link("external", "license"); !strings.HasPrefix(got, "https://") {
		t.Errorf("license link = %q, want an https URL", got)
	}
	if got := cfg.Link("external", "license"); strings.ContainsAny(got, "<>") {
		t.Errorf("license link = %q still contains a placeholder", got)
	}
	if got := cfg.Link("inApp", "terms"); !strings.HasPrefix(got, "/") {
		t.Errorf("in-app terms route = %q, want an app-relative route", got)
	}
	if got := cfg.Asset("mark"); got == "" {
		t.Error("Asset(\"mark\") is empty")
	}
}

func TestTextResolvesSharedCopy(t *testing.T) {
	cfg := Load()
	for _, locale := range []string{"zh-CN", "en-US"} {
		if got := cfg.Text("termsBody", locale); got == "" {
			t.Errorf("Text(termsBody, %q) is empty", locale)
		}
	}
	if got := cfg.Text("no-such-key", "en-US"); got != "" {
		t.Errorf("Text on an unknown key = %q, want empty", got)
	}
}

func TestLocalizeFallbackOrder(t *testing.T) {
	for _, tc := range []struct {
		name   string
		values Localized
		locale string
		want   string
	}{
		{"exact match wins", Localized{"zh-CN": "精确", "en-US": "exact"}, "zh-CN", "精确"},
		{"language subtag match", Localized{"zh-CN": "简体", "en-US": "en"}, "zh-Hant", "简体"},
		{"english default", Localized{"en-US": "en"}, "ja-JP", "en"},
		{"empty value skipped", Localized{"zh-CN": "", "en-US": "en"}, "zh-CN", "en"},
		{"any value as last resort", Localized{"ja-JP": "日本語"}, "ko-KR", "日本語"},
		{"nil map", nil, "en-US", ""},
		{"all empty", Localized{"en-US": ""}, "en-US", ""},
	} {
		if got := localize(tc.values, tc.locale); got != tc.want {
			t.Errorf("%s: localize(%v, %q) = %q, want %q", tc.name, tc.values, tc.locale, got, tc.want)
		}
	}
}

File: web/src/lib/brand.test.ts
import { describe, expect, it } from 'vitest'

import {
  BRAND_SCHEMA_VERSION,
  IDENTIFIER,
  brand,
  brandAsset,
  brandColor,
  brandCopyright,
  brandCopyrightFor,
  brandDisplay,
  brandIdentifier,
  brandLink,
  brandName,
  brandShortName,
  brandTeamName,
  brandText,
  localize,
} from './brand'

describe('brand configuration', () => {
  it('embeds the schema version this module understands', () => {
    expect(brand.schemaVersion).toBe(BRAND_SCHEMA_VERSION)
  })

  it('resolves the full and short names per locale', () => {
    expect(brandName('zh-CN')).toBe('布丁盒子')
    expect(brandName('en-US')).toBe('Pudding Box')
    expect(brandShortName('zh-CN')).toBe('布丁盒子')
    expect(brandShortName('en-US')).toBe('Pudding')
  })

  // The regression this module exists to prevent: the UI locale store holds
  // bare 'zh'/'en', the configuration is keyed 'zh-CN'/'en-US'. Indexing the
  // map directly renders English text inside the Chinese UI.
  it('resolves the bare locale the UI store actually holds', () => {
    expect(brandName('zh')).toBe('布丁盒子')
    expect(brandName('en')).toBe('Pudding Box')
    expect(brandTeamName('zh')).toBe('布丁盒子工作室')
    expect(brandTeamName('en')).toBe('Pudding Box Studio')
  })

  it('never returns an empty string for an unknown locale', () => {
    for (const locale of ['', 'de-DE', 'ja', 'zh-Hant']) {
      expect(brandName(locale)).not.toBe('')
      expect(brandTeamName(locale)).not.toBe('')
    }
  })
})

describe('localize', () => {
  it('prefers an exact tag', () => {
    expect(localize({ 'zh-CN': '精确', 'en-US': 'exact' }, 'zh-CN')).toBe('精确')
  })

  it('falls back through the language subtag', () => {
    expect(localize({ 'zh-CN': '简体', 'en-US': 'en' }, 'zh-Hant')).toBe('简体')
  })

  it('falls back to English, then to any value', () => {
    expect(localize({ 'en-US': 'en' }, 'ja-JP')).toBe('en')
    expect(localize({ 'ja-JP': '日本語' }, 'ko-KR')).toBe('日本語')
  })

  it('skips empty values instead of returning a blank string', () => {
    expect(localize({ 'zh-CN': '', 'en-US': 'en' }, 'zh-CN')).toBe('en')
  })

  it('tolerates a missing map', () => {
    expect(localize(undefined, 'en-US')).toBe('')
    expect(localize({}, 'en-US')).toBe('')
  })
})

describe('copyright', () => {
  it('keeps the placeholder in the raw value', () => {
    expect(brandCopyright('zh-CN')).toContain('{year}')
  })

  it('substitutes an explicit year', () => {
    expect(brandCopyrightFor('zh-CN', 2026)).toBe('© 2026 布丁盒子工作室')
    expect(brandCopyrightFor('en-US', 2026)).toBe('© 2026 Pudding Box Studio')
  })

  it('defaults to the current year and leaves no placeholder behind', () => {
    const rendered = brandCopyrightFor('zh')
    expect(rendered).not.toContain('{year}')
    expect(rendered).toContain(String(new Date().getFullYear()))
  })
})

describe('class B fixed display values', () => {
  it('are single values, not locale maps', () => {
    expect(brandDisplay('windows', 'productName')).toBe('Pudding Box')
    expect(brandDisplay('windows', 'fileDescription')).toBe('布丁盒子')
  })

  it('return empty rather than throwing on an unknown group', () => {
    expect(brandDisplay('nope', 'productName')).toBe('')
  })
})

describe('class C identifiers and paths', () => {
  it('resolves every declared key', () => {
    for (const key of Object.values(IDENTIFIER)) {
      expect(brandIdentifier(key), `identifier ${key}`).not.toBe('')
    }
  })

  // Class C is machine-read: a Chinese character or a space in a path, a lock
  // name or a protocol scheme is a functional bug, not a cosmetic one.
  it('are printable ASCII without spaces', () => {
    const groups: Record<string, Record<string, string>> = {
      'identifiers.current': brand.identifiers.current,
      'identifiers.future': brand.identifiers.future,
      'visual.logo': brand.visual.logo,
      // Absent until the brand colour is agreed; an empty group contributes
      // nothing rather than throwing.
      'visual.colors': brand.visual.colors ?? {},
    }
    for (const [group, links] of Object.entries(brand.links)) groups[`links.${group}`] = links

    for (const [group, values] of Object.entries(groups)) {
      for (const [key, value] of Object.entries(values)) {
        expect(value, `${group}.${key}`).toMatch(/^[\x21-\x7E]+$/)
      }
    }
  })

  it('keeps the single-instance lock off the upstream identifier', () => {
    expect(brandIdentifier(IDENTIFIER.singleInstanceId)).toBe('app.puddingbox.desktop')
  })

  // These stay on the upstream values this phase; changing them breaks
  // in-place upgrades for already-installed copies.
  it('keeps the compatibility identifiers unchanged', () => {
    expect(brandIdentifier(IDENTIFIER.cliCommand)).toBe('octo')
    expect(brandIdentifier(IDENTIFIER.configDir)).toBe('~/.octo')
    expect(brandIdentifier(IDENTIFIER.envPrefix)).toBe('OCTO_')
  })

  it('exposes usable links, assets and colours', () => {
    expect(brandLink('external', 'license')).toMatch(/^https:\/\//)
    expect(brandLink('external', 'license')).not.toMatch(/[<>]/)
    expect(brandLink('inApp', 'terms')).toMatch(/^\//)
    expect(brandLink('nope', 'license')).toBe('')
    expect(brandAsset('mark')).not.toBe('')
    // The primary brand colour is decided (sampled from logo-mark.png,
    // 2026-09-09); an unknown colour group must still degrade to an empty
    // string rather than throw on the missing key.
    expect(brandColor('primary')).toBe('#437EB1')
    expect(brandColor('nope')).toBe('')
  })
})

describe('shared copy', () => {
  it('resolves both locales', () => {
    expect(brandText('termsBody', 'zh-CN')).not.toBe('')
    expect(brandText('termsBody', 'en-US')).not.toBe('')
  })

  it('returns empty for an unknown key', () => {
    expect(brandText('no-such-key', 'en-US')).toBe('')
  })
})
